### ISSUE TITLE
Extend precaching to target non-precache caches and use custom strategies

### ISSUE THREAD

**Issue by westonruter** [GoogleChrome/workbox#1612]
**Library Affected**:
*workbox-precaching*

**Browser & Platform**:
*all browsers*

**Feature Request Description**:

In looking at @sebastianbenz's [minimal service worker for websites built with AMP](https://github.com/ampproject/amp-by-example/blob/master/boilerplate-generator/templates/files/serviceworkerJs.js) there is the following code:

```js
//Pre-cache the AMP Runtime
self.addEventListener('install', event => {
  const urls = [
    'https://cdn.ampproject.org/v0.js',
    // Add AMP extensions used on your pages
    // Add fonts, icons, logos used on your pages
  ];
  /* ... */
  event.waitUntil(
    caches.open(workbox.core.cacheNames.runtime).then(cache => cache.addAll(urls))
  );
});

/* ... */

// Serve the AMP Runtime from cache and check for an updated version in the background
workbox.routing.registerRoute(
  /https:\/\/cdn\.ampproject\.org\/.*/,
  workbox.strategies.staleWhileRevalidate()
);
```

I was curious why `workbox.precaching.precacheAndRoute()` wasn't being used, but there seem to be two reasons:

1. A strategy other than `cacheFirst` is needed, namely `staleWhileRevalidate`.
2. The precached URLs need to be stored in the runtime cache for the `staleWhileRevalidate` strategy to locate.

That got me to thinking, ¿what if `workbox.precaching.precache()` could be used instead? Consider this alternative:

```js
//Pre-cache the AMP Runtime
workbox.precaching.precache([
    { 
        url: 'https://cdn.ampproject.org/v0.js', 
        cacheName: workbox.core.cacheNames.runtime
    }
    // Add AMP extensions used on your pages
    // Add fonts, icons, logos used on your pages
]);

// Serve the AMP Runtime from cache and check for an updated version in the background
workbox.routing.registerRoute(
  /https:\/\/cdn\.ampproject\.org\/.*/,
  workbox.strategies.staleWhileRevalidate()
);
```

Now, in this example the difference is not that great. But what if the “runtime-precached” URLs could have a `revision` associated with each of them? It doesn't seem possible currently to use the `precaching` utilities to cache URLs in any cache other than `workbox.core.cacheNames.precache`.

One more missing piece here is what happens when someone calls `workbox.precaching.addRoute()`. Since it adds a cache-first strategy, it will prevent the desired stale-while-revalidate behavior. So what if the strategy could be specified along with each precache entry as opposed to one strategy for all? For example:

```js
//Pre-cache the AMP Runtime
workbox.precaching.precacheAndRoute([
    { 
        url: 'https://cdn.ampproject.org/v0.js',
        cacheName: workbox.core.cacheNames.runtime,
        handler: workbox.strategies.staleWhileRevalidate(),
        revision: '1535039767807'
    }
]);
```

For any precache entries that lack a `handler`, the default cache-first strategy would be done.

----

The context for this is looking at how to best to implement the minimal AMP service worker in the AMP WordPress plugin: https://github.com/Automattic/amp-wp/pull/1261

This AMP plugin pull request in turn depends on another pull request in the PWA feature plugin which is exploring how to best provide a PHP abstraction for the Workbox JavaScript APIs: https://github.com/xwp/pwa-wp/pull/53

From the AMP plugin pull request, the PHP API currently looks like this:

```php
wp_service_workers()->register_precached_route( 'https://cdn.ampproject.org/v0.js', array(
	'cache' => WP_Service_Workers::RUNTIME_CACHE_NAME,
) );
wp_service_workers()->register_cached_route(
	'^https:\/\/cdn\.ampproject\.org\/.*',
	WP_Service_Workers::STRATEGY_STALE_WHILE_REVALIDATE
);
```

The `wp_service_workers()->register_precached_route()` PHP function here [basically implements](https://github.com/xwp/pwa-wp/blob/b56ed3dd4fcd72aea44040b4fc6a7d5e6b3c619d/wp-includes/js/service-worker-precaching.js) what I proposed above in Workbox itself, but without any handling of the `revision` for non-precache caches. So this same PHP method can be used as an abstraction for `workbox.precaching.precache()` and manually precaching to another cache.

**Comment by jeffposnick** [comment]
Hey @westonruter!

We would ideally like to keep `workbox-precaching` focused on ingesting manifest entries that are derived from local files available at build time, rather than arbitrary, third-party URLs. That keeps the mental model of how the entries are served (always cache-first) and updated (always as the result of a revision or URL update in the manifest) consistent.

Using runtime caching for third-party assets which we can't version at build time is a model that tends to work, but it can get a bit awkward when you also need to prime the cache ahead of time. I might be missing something crucial about the WordPress use case, but I *think* what your suggestion boils down to is something that would make it easier to prime the cache at `install`-time, without waiting for the cache to be populated after the first runtime request is handled. It sounds to me as being similar to https://github.com/GoogleChrome/workbox/issues/368. I think pursuing this sort of thing as being separate from `workbox-precaching` is worthwhile.

Or am I missing something subtle here about your request?

**Comment by westonruter** [comment]
@jeffposnick thanks for the feedback. I suppose what I'm getting at is whether Workbox should do anything to facilitate what @sebastianbenz has [done here](https://github.com/ampproject/amp-by-example/blob/master/boilerplate-generator/templates/files/serviceworkerJs.js#L9-L22):

```js
// Pre-cache the AMP Runtime
self.addEventListener('install', event => {
  const urls = [
    'https://cdn.ampproject.org/v0.js',
    // Add AMP extensions used on your pages
    // Add fonts, icons, logos used on your pages
  ];
  if (OFFLINE_PAGE) {
    urls.push(OFFLINE_PAGE);
  }
  event.waitUntil(
      caches.open(workbox.core.cacheNames.runtime).then(cache => cache.addAll(urls))
  );
});
```

(Actually, it seems `OFFLINE_PAGE` would perhaps be better suited for the precache than this runtime cache at `install`?)

In short: I'm curious about how common of a use case this is.

If it is a common use case, then I should think the WordPress service worker API for managing cached assets should directly support it with a PHP interface. And if so, then perhaps Workbox itself should have some additional `precaching` facility for it. But if it is not common or general enough to warrant a PHP API, then it can be implemented in WordPress by registering raw JS code (as above) to be included in the service worker to do the same.
